tracing-core/src/callsite.rs

```diff
@@ -169,3 +177,169 @@ impl Hash for Identifier {
         (self.0 as *const dyn Callsite).hash(state)
     }
 }
+
+/// An intrusive atomic push only linked-list.
+struct LinkedList {
+    head: AtomicPtr<Registration>,
+}
+
+unsafe impl Send for LinkedList {}
+unsafe impl Sync for LinkedList {}
+
+/// Represents a registration within the callsite cache.
+///
+/// This type enables the intrusive aspect of the callsite cache by
+/// avoiding allocations and letting the callsite's themselves store
+/// the next callsite in the cache.
```

Comment:
i don't think it's necessary for the docs to explain these implementation details; i'd prefer this focus on what the user is expected to do with this type (e.g. every type implementing `Callsite` needs to own its own `Registration` that points back to it...)